CONTRACTS: Allow pure contracts.

remi-delmas-3000 opened the pull request:
A pure contract is modelled as a function declaration with a contract. 

```
int foo(args)
__CPROVER_requires(R)
__CPROVER_ensures(E)
__CPROVER_assigns(A)
;
```

We reject multiple declarations for such functions (with or without a body).

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [n/a] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

tautschnig commented on regression/contracts/named-contracts/main-fail2.c:
```diff
@@ -0,0 +1,25 @@
+{
+  int retval = foo(&arr, 10);
+  return 0;
+}
```
GitHub says there is a missing newline.

tautschnig commented on regression/contracts/named-contracts/main-pass.c:
```diff
@@ -0,0 +1,32 @@
+  assert(arr[8] == 2);
+  assert(arr[9] == 0);
+  return 0;
+}
```
GitHub says there is a missing newline.

tautschnig commented on regression/contracts/named-contracts/test-fail1.desc:
```diff
@@ -0,0 +1,10 @@
+CORE
+main-fail1.c
+
+^main-fail1.c.* error: the contract for function 'foo' must be attached to the first declaration.$
```
This error message sounds suspicious when it says "first declaration": there should not be a particular order of declarations required, unless this really is restricting to single translation units? What happens when linking declarations (and perhaps: definitions) from multiple translation units?

remi-delmas-3000 replied:
An order seems to be implied by the existence of `typecheck_new_symbol` and `typecheck_redefinition_non_type` which fuses an old definition with a new one. I don't know how this maps to source order or linking, but I tried reject redefinition when the first that was seen was a pure contract (i.e. a function with contract clauses but no body). If linking calls `typcheck_redefinition_non_type` for all redefinitions this should ensure the same invariant.

tautschnig commented on regression/contracts/named-contracts/test-fail2.desc:
```diff
@@ -0,0 +1,10 @@
+^EXIT=(1|64)$
+^SIGNAL=0$
+--
+This test checks that if a contract was attached to a functio declaration
```
s/functio/function/

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -1,3 +1,4 @@
+#include <iostream>
```
Debugging left-over?

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -148,11 +149,16 @@ void c_typecheck_baset::typecheck_new_symbol(symbolt &symbol)
   {
-    if(symbol.value.is_not_nil() &&
-       !symbol.is_macro)
+    if(symbol.value.is_not_nil() && !symbol.is_macro)
```
Unnecessary change?

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -395,10 +393,23 @@ void c_typecheck_baset::typecheck_redefinition_non_type(
+        {
+          error().source_location = new_symbol.location;
+          error()
+            << "function '" << new_symbol.display_name()
+            << "' cannot have a declaration with a contract and a definition."
```
I'm struggling to understand what this means: is it not permitted to have a function contract for a function that also has a definition? Or is it just that the definition must not have a contract attached to it? (But then I don't see where this is being checked.)

remi-delmas-3000 replied:
The invariant I'm trying to enforce is that if there are multiple declarations, and one of these declarations is a pure contract, we have an error. It does not prevent to use the existing contract style where you can have a declaration and then a definition with a body and a contract.

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -364,15 +370,7 @@ void c_typecheck_baset::typecheck_redefinition_non_type(
-            if(p_s_it!=symbol_table.symbols.end())
-              symbol_table.erase(p_s_it);
-          }
+          remove_parameters_from_symbol_table(old_symbol);
```
I'd prefer to see this refactoring happen in a commit of its own.

remi-delmas-3000 replied:
I split the PR in three commits, the first one is the refactoring commit.

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -499,56 +520,82 @@ void c_typecheck_baset::typecheck_redefinition_non_type(
     p_symbol.type = p.type();
-    p_symbol.name=identifier;
-    p_symbol.base_name=base_name;
+    p_symbol.name = identifier;
+    p_symbol.base_name = base_name;
```
Unnecessary change.

remi-delmas-3000 replied:
this is to make clang-format happy

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -499,56 +520,82 @@ void c_typecheck_baset::typecheck_redefinition_non_type(
     // produce identifier
     irep_idt base_name = p.get_base_name();
-    irep_idt identifier=id2string(symbol.name)+"::"+id2string(base_name);
+    irep_idt identifier = id2string(symbol.name) + "::" + id2string(base_name);
```
Unnecessary change.

remi-delmas-3000 replied:
this is to make clang-format happy

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -499,56 +520,82 @@ void c_typecheck_baset::typecheck_redefinition_non_type(
     if(p.get_base_name().empty())
     {
-      irep_idt base_name="#anon"+std::to_string(anon_counter++);
+      irep_idt base_name = "#anon" + std::to_string(anon_counter++);
```
Unnecessary change.

remi-delmas-3000 replied:
this is to make clang-format happy

tautschnig reviewed (changes requested):
I'm afraid this needs some house keeping to make it easier to properly review: whitespace changes result in a rather weird diff being presented. Also, any refactoring should be in a commit of its own to make it easier to understand what the actual changes in behaviour are.

Commit:
CONTRACTS: allow pure contracts.

A pure contrat is a function
declaration without body and
with contract clauses.

Commit:
CONTRACTS: reject redeclaration of pure contracts

We reject redeclarations of a function
if one of these declarations is in
pure contract form.

Commit:
Refactoring: adding `add_parameters_to_symbol_table` for code reuse
